### PR TITLE
Reload themes after authentication

### LINKED ISSUE #389: Rechargement du catalogue après login
Vu qu'un utilisateur peut avoir le droit de voir certaines couches uniquement quand il est logué, il faut recharger le catalogue après login

### PR DESCRIPTION
**Pull request by elemoine**
The objective of this PR is to reload themes after the user has logged in or logged out. This is still WIP for now.

Fixes #389.

### CODE REVIEW

**Comment by elemoine** [comment]
This is now ready for review.

With no LDAP access I haven't been able to fully test that PR. I think it would good if someone from the Luxembourg could test my changes before we merge that PR.

When an anonymous user loads the application in the browser there should one request to the `themes` service and only one.

When an already authenticated user loads the application in the browser there should be two requests to the `themes` service. The first one is done eagerly, because, as discussed with @jaykayone in #389, in most cases, users are not yet logged in when they load the application.

When the user logs in there should be one request to the `themes` service, and the `themes` service should see an "authenticated" user.

When the user logs out there should be one request to the `themes` service, and the `themes` service should see an "anonymous" user.

Please test and review.

**Comment by rmichaelis** [comment]
I can't test.
The build fails due to the proxy error due to the git protocol.
I still don't know how to bypass this problem.

**Comment by elemoine** [comment]
Could you please test with patch:

```diff
diff --git a/CONST_requirements.txt b/CONST_requirements.txt
index 329b2c4..f229fc2 100644
--- a/CONST_requirements.txt
+++ b/CONST_requirements.txt
@@ -3,5 +3,5 @@
 --find-links http://pypi.camptocamp.net/internal-pypi/index/c2cgeoportal-win
 -r CONST_versions.txt
 -e git+https://github.com/camptocamp/pyramid_closure@819bc43420b3cd924d8698c5a9606592c19dbb15#egg=pyramid_closure
--e git+https://github.com/Pylons/pyramid@1e02bbfc0df09259bf207112acf019c8dba44a90#egg=pyramid
+https://github.com/Pylons/pyramid/archive/1e02bbfc0df09259bf207112acf019c8dba44a90.zip#egg=pyramid
 -e .
```

**Comment by rmichaelis** [comment]
Thanks for the patch. It woks fine. 
I'm now able to test the PR.

**Comment by elemoine** [comment]
> Thanks for the patch. It woks fine.
> I'm now able to test the PR.

So I'll create a PR with this. Thanks for testing.

**Comment by elemoine** [comment]
@rmichaelis I now set `cache` to `false` when reloading themes with `$http.get`. I hope this will fix the issue. Can you please test again?

**Comment by elemoine** [comment]
As discussed with @sbrunner we should use a `role` parameter in the URL.

**Comment by elemoine** [comment]
Patch updated.

So, when the role changes the themes are reloaded, and the role id is sent to the `themes` service using `&role=<role_id>` in the query string.

@sbrunner, can you please confirm that this is what you had in mind?

**Comment by elemoine** [comment]
Has this been tested?

**Comment by rmichaelis** [comment]
Sorry. Still not tested. 
I do it today.

**Comment by rmichaelis** [comment]
I think it's not working. 
After page reload, themes are loaded. 
Then I connect and the themes webservice is not called. 
Then disconnect and the themes webservice is still not called.  

capture

**Comment by elemoine** [comment]
Argh.

I've only done tests in the dev tools' console:

```js
> $(document.body).scope().mainCtrl.roleId = 2;  // change the roleId in the main controller
> $(document.body).scope().$digest();  // trigger a digest
```

And this does trigger a new request to the `themes` service.

So the problem may be in the communication between the authentication directive and the main controller.

**Comment by elemoine** [comment]
There was a typo in `index.html`. I had

```html
<app-authentication app-authentication-roleid="mainCtrl.rodeId"></app-authentication>
```

instead of 

```html
<app-authentication app-authentication-roleid="mainCtrl.roleId"></app-authentication>
```

It should work now. At least a request to the `themes` service should happen when the user logs in or logs out.

File: geoportailv3/static/js/authentication/authenticationdirective.js [hunk 1]
```diff
@@ -207,17 +213,26 @@ app.AuthenticationController.prototype.hasError = function() {
...
+ * @param {string|undefined} mail Mail.
+ * @param {string|undefined} name Name.
  */
 app.AuthenticationController.prototype.setUserInfo = function(
-    login, role, role_id, mail, name) {
+    login, role, roleId, mail, name) {
   this['login'] = login;
   this['role'] = role;
-  this['role_id'] = role_id;
+  this['rodeId'] = roleId;
```
Comment (rmichaelis) [inline geoportailv3/static/js/authentication/authenticationdirective.js]: 'roleId' and not 'rodeId'

Comment (elemoine) [inline geoportailv3/static/js/authentication/authenticationdirective.js]: Woa! I really have a problem with this word. Thanks for catching this.

**Comment by elemoine** [comment]
Thanks @rmichaelis. I fixed the remaining typo.

**Comment by rmichaelis** [comment]
The functionality is ok.  
When I login and log out the theme is correctly reloaded, and the new layers are shown in catalog.

But the cache problem is still remaining. 
If a user has roleId = 1. 
Each response from the webservice using parameter roleId=1 is coming from the cache and not from the network.
That means, that the catalog is not updated with the last state of the database.

**Comment by elemoine** [comment]
> But the cache problem is still remaining.
> If a user has roleId = 1.
> Each response from the webservice using parameter roleId=1 is coming from the cache and not from the network.
> That means, that the catalog is not updated with the last state of the database.

Can you confirm that the request to the `themes` web service includes a `role` parameter (whose value is the role id) when the user is logged in?

**Comment by rmichaelis** [comment]
>Can you confirm that the request to the `themes` web service includes a `role` parameter (whose value is the role id) when the user is logged in?

Yes I confirm. It's working as expected. Except that only the first request of each roleId is coming from network. The next ones are coming from the cache.

**Comment by elemoine** [comment]
> Yes I confirm. It's working as expected. Except that only the first request of each roleId is coming from network. The next ones are coming from the cache.

I understand. What cache are you talking about here?

**Comment by rmichaelis** [comment]
>I understand. What cache are you talking about here?

I suspect the angular cache, because using the private mode of Firefox doesn't fix the problem.

**Comment by elemoine** [comment]
What do you think about https://github.com/Geoportail-Luxembourg/geoportailv3/commit/f417d5538dbcdb519c406fc41e4d8c6cd7634918?

**Comment by rmichaelis** [comment]
No, it doesn't work. Still the same behavior.

**Comment by elemoine** [comment]
> No, it doesn't work. Still the same behavior.

That sucks. I will need to study this in more detail. Thanks a lot for testing!

**Comment by rmichaelis** [comment]
It seems that responses are coming from BFCache
cache control is actually set to "max-age=864000, private"

**Comment by elemoine** [comment]
> It seems that responses are coming from BFCache
> cache control is actually set to "max-age=864000, private"

@sbrunner, `themes` responses should not be cacheable, right?

**Comment by jaykayone** [comment]
what about adding a random parameter?

**Comment by sbrunner** [comment]
@elemoine , no it should, it's why we should add the role parameter to have one cache per role.
Then we can do many login / logout with only two request :-)

**Comment by rmichaelis** [comment]
If the list of layers changed in webservice, how could the user have the last results ? 
Is a page reload enough ?

**Comment by sbrunner** [comment]
Yes because we will also add a cache version :-)

**Comment by jaykayone** [comment]
i have not understood your answer @sbrunner

**Comment by elemoine** [comment]
> i have not understood your answer @sbrunner

@sbrunner is saying that, with https://github.com/camptocamp/c2cgeoportal/pull/1328, `themes` requests will include a `cache_version` parameter, whose value will change when the WSGI application cache is invalidated.

@rmichaelis, this also means that a page reload will be necessary to see changes in the database.

**Comment by jaykayone** [comment]
OK for me

**Comment by elemoine** [comment]
May I suggest to merge this and sort out the caching issues with separate GitHub issues/pull requests?

**Comment by elemoine** [comment]
Do people agree with merging this?

**Comment by rmichaelis** [comment]
Yes, I agree.

**Pull request merged**